fix(node): flaky crypto generate signing key spec

ffakenz opened the pull request:
Fixes #378

ffakenz committed:
Fix flaky spec by updating `generateSigningKey` to add a left padding.

Using length of seed as value for the left pad.
This will ensure 2 seeds of different size will get different results,
so cases like ("", "\NUL") returning same output will not ocurr.
This will also avoid cases where a seed is equal to the nth padded version of another,
for example: ("1\NUL","1\NUL\NUL")

ffakenz committed:
Fix signing and verification keys hardcoded tests.

ffakenz committed:
Add failing test: case of seed exceeding max required size for the signing algorithm.

ffakenz committed:
Fix broken test for seeds exceeding max size of required by signing algorithm

First draft attempt using hash of seed instead of length.

ffakenz committed:
Fix calculation for hash of seed.

We were `casting` to Hash type not calculating a hash value.

ffakenz committed:
Remove unused import

ffakenz committed:
Fix hardcoded test for signing key

ffakenz committed:
Fix hardcoded test for verification key

ch1bo commented on hydra-node/src/Hydra/Crypto.hs:
```diff
@@ -82,8 +82,9 @@ generateSigningKey seed =
-  padded = seed <> BS.pack (replicate (needed - provided) 0)
+  hashOfSeed = digest (Proxy :: Proxy SHA256) seed
+  provided = BS.length hashOfSeed + BS.length seed
+  padded = hashOfSeed <> seed <> BS.pack (replicate (needed - provided) 0)
```
If we use the hash of the seed, we do not need to do any padding.

ch1bo reviewed (changes requested):
Should: Simplify key derivation even further to only use the hash and not the seed directly as well.

ffakenz committed:
Add missing new line at EOF

ffakenz committed:
Add missing imports

ffakenz committed:
Fix hardcoded tui e2e spec

ffakenz committed:
Remove unnecessary import

ch1bo commented on hydra-node/src/Hydra/Crypto.hs:
```diff
@@ -180,9 +180,8 @@ generateSigningKey :: ByteString -> SigningKey HydraKey
-  provided = BS.length seed
-  padded = seed <> BS.pack (replicate (needed - provided) 0)
+  hashOfSeed = digest (Proxy :: Proxy SHA256) seed
+  padded = hashOfSeed <> seed
```
The derived key can only be the hash. No need to suffix the seed. Also, the variable name `padded` does not make sense anymore. Just use `hashOfSeed` instead of it.

ffakenz committed:
Fix hardcoded crypto spec for verification key

ffakenz committed:
Use only hash of seed as derived key.

ch1bo commented on hydra-node/test/Hydra/CryptoSpec.hs:
```diff
@@ -24,7 +24,11 @@ specSigningKey :: Spec
+      show (generateSigningKey "aaa") `shouldContain` "\"983487"
+    it "can be generated when seed exceeds the max seed size for algorithm" $
+      let exceedingSizeSeedA = show $ replicate 32 'x' <> "a"
+          exceedingSizeSeedB = show $ replicate 32 'x' <> "b"
```
`show` should not be needed here. Usually it will wrap a bytestring into quotes. As this is likely the overloaded version from relude (https://hackage.haskell.org/package/relude-1.1.0.0/docs/Relude-String-Conversion.html#v:show), it will use the `IsString` of `ByteString`. That likely is `Data.ByteString.Char8.pack`. The extra `"` in the beginning and end will still result in 2 + 32 + 1 length seeds.. which is still okay, but not exactly what you intended to write here.

Should: Don't use show here, as it will include extra `"`

ch1bo reviewed (approved):
Consider removing the `show`, but otherwise looks good to me.

ffakenz committed:
Fix exceeding size spec by using Char8.pack instead of show

The use of `show` was introducing extra " inside the ByteString,
and that is not exactly what we want in order to reproduce precisely the
scenario.
